### ISSUE TITLE
Planned dynamic string API

### ISSUE THREAD

**Issue by JacksonAllan** [JacksonAllan/CC#12]
As I'm currently in the process of adding the final planned container—dynamic, null-terminated strings—to CC, I've posted a call for public feedback on the API [here](https://old.reddit.com/r/C_Programming/comments/1epszvi/seeking_feedback_on_dynamic_string_api/). So if any current CC users have thoughts on the matter, please feel free to respond here or in the linked Reddit post. Much appreciated!

**Comment by gabewillen** [comment]
@JacksonAllan 

The string API looks great overall and aligns nicely with the other containers. I have just one piece of feedback, which is more of a personal preference regarding a method name.

I would suggest renaming:

`size_t cap(str(el_ty) *cntr)`

to:

`size_t capacity(str(el_ty) *cntr)`

I prefer more explicit names, and since the rest of the API (except for `init`) uses full words, this would be consistent. Additionally, at a glance, someone might mistakenly assume that cap refers to “capitalize” rather than “capacity,” especially without seeing a type declaration.

As I mentioned, this is just a personal opinion, so feel free to take it as you see fit.

Gabe

**Comment by JacksonAllan** [comment]
Hi Gabe. Thanks very much for taking the time to provide feedback!

I agree that `cap` should ideally be `capacity`. Likewise, `prev` should perhaps be `previous`. There are a few other abbreviated API macros, namely `init`, `r_end`, `r_for_each`, and `erase_itr`. But in those cases, adopting the full-word equivalent would result in a much longer identifier (ten or more characters), so the argument for changing them is, I think, weaker.

Another issue with `cap` is that `cap`/`capacity` is a misnomer when applied to `map` and `set` because the macro "returns" the bucket count, not the number of elements the container could hold without reallocation. Currently, this discrepancy is addressed by a rather clumsy comment in the documentation: `Note that the number of elements a map can accommodate without rehashing is not its capacity but its capacity multiplied by the max load factor associated with its key type`. In C++, `std::unordered_map` instead names its capacity-like function `bucket_count`. But in CC, every new API macro becomes a reserved global identifier (at least for function and macro names), so there is a strong incentive not to add any more API macros. When designing the API, I wasn't able to think of any name that would accurately describe both the capacity of a vector and the bucket count of a hash table, so I settled on using `cap` for both as the lesser of two evils (i.e. I thought having a slight misnomer in the case of maps and sets is better than reserving yet another identifier).

Of course, changing `cap` to `capacity` will be a breaking API change and should be done is phases, with `cap` being deprecated first and fully removed later. I think that's okay as there is already a breaking API change on the horizon: `prev` and `r_end` have already been deprecated for maps and sets and are waiting for removal in the next major version. So if I do change `cap` (and possibly `prev`), I'll try to time the change so that the old identifiers are fully removed at the same time as those two `map` and `set` operations are removed.

**Comment by JacksonAllan** [comment]
Also, I forgot to mention that I've long felt that `push` and `push_n` should instead have been `append` and `append_n`. The name `push` is obviously inspired by C++'s `push_back`. But without that additional context, the name isn't very meaningful, and not all C programmers are necessarily well-versed in C++. So that's another name that we perhaps ought to reconsider if we're going to revamp the API.

**Comment by gabewillen** [comment]
I agree with all of the above and I always prefer explicitly long "document by code" style naming conventions. A few extra key strokes initially beats having to follow source links to find out what variable `c` type / assignment is somewhere.

**Comment by JacksonAllan** [comment]
One more thought on API-macro names that I'll post here so that I don't forget it later: `cleanup` should ideally be `clean_up`, in keeping with the other names.

**Comment by JacksonAllan** [comment]
CC strings are now available via the [v1.4.0 Prerelease](https://github.com/JacksonAllan/CC/releases/tag/v1.4.0_Prerelease). A few important differences between the API that I outlined earlier and the final implementation:
* In the original plan and prototype, `push` and `insert` were used to insert a variadic list of generic, formatted arguments into a string. In the final API, this functionality is instead provided by the dedicated macros `push_fmt` and `insert_fmt`. `push` and `insert`, on the other hand, insert individual elements/characters, in keeping with the way those macros behave with other containers.
* The final API supports [heterogeneous insertion and lookup](https://github.com/JacksonAllan/CC/tree/v1.4.0_prerelease#string-interoperability) for containers that use CC strings as their key and/or element type. The main purpose of this feature is to simplify the use of such containers with regular C strings. It also allows CC strings to easily be *copied* into such containers (e.g. by inserting `first( &our_str )`  instead of `our_str`), rather than inserted in the regular manner that transfers ownership to the container.

**Comment by JacksonAllan** [comment]
Strings have now been introduced with [v1.4.0](https://github.com/JacksonAllan/CC/releases/tag/v1.4.0). Changes to API macro names are due for the next major release.
